### PR TITLE
WIP: Requirement lists

### PR DESCRIPTION
**Pull request by hikinggrass**

### CODE REVIEW

**Comment by caller** [comment]
@tmolitor-stud-tu  you last commit wasn't signed, so we cant merge this... can you fix that?

**Comment by tmolitor-stud-tu** [comment]
It is still WIP (the C++ code generator stuff is still missing), so nothing to merge yet

File: everestjs/utils.hpp
Comment (a-w50) [inline everestjs/utils.hpp:20]: What is the reason for calling BOOST_THROW_EXCEPTION and catching in again afterwards? Is it only to get the file, line and function information via boost log?

File: lib/config.cpp
Comment (a-w50) [inline lib/config.cpp:773]: why not use the previously defined variable, ie main.contains(connection_module_id) ?

File: lib/config.cpp
Comment (a-w50) [inline lib/config.cpp:767]: For the debug message it would be helpful to also see the actual value of connections.size()

File: lib/config.cpp
Comment (a-w50) [inline lib/config.cpp:777]: would rather write "referenced in connection"

File: lib/config.cpp
Comment (a-w50) [inline lib/config.cpp:806]: we should definitely switch to libfmt some day

File: include/framework/everest.hpp
Comment (a-w50) [inline include/framework/everest.hpp:78]: uint64_t? That's what I call "think large"

File: everestjs/everestjs.cpp
Comment (a-w50) [inline everestjs/everestjs.cpp:123]: good find :)

File: everestjs/everestjs.cpp
Comment (a-w50) [inline everestjs/everestjs.cpp:399]: what's the difference to the auto requirement_impl_intf = config->get_interfaces()[requirement_module_name][requirement_impl_id] from before?

File: src/main.cpp
Comment (a-w50) [inline src/main.cpp:151]: :bowtie: for renaming params

File: include/framework/everest.hpp
Comment (a-w50) [inline include/framework/everest.hpp:92]: I wonder if we need these 2 subscribe_var overloads at all? The one with the ValueCallback should be sufficient I guess, because we shouldn't pass raw json - and if so, the raw json can also be put into boost::any

File: lib/config.cpp
Comment (a-w50) [inline lib/config.cpp:81]: instead of !config_map.contains(config_entry_name), I would prefer to check for config_entry_value == null

File: lib/config.cpp
Comment (a-w50) [inline lib/config.cpp:82]: Do we need this at all, because below we're using the patch

File: lib/config.cpp
Comment (a-w50) [inline lib/config.cpp:181]: It's good to have these patching process in all places, I wonder if this will change/break something because we haven't had it before.  There is no strict need for it, right?  For future I definitely would like to keep it

File: lib/config.cpp
Comment (a-w50) [inline lib/config.cpp:187]: Why did you drop the  if (this->manifests[module_name].contains("config"))?
It is not required to be existent, and below it is accessed, so this might lead to a null reference

File: lib/config.cpp
Comment (a-w50) [inline lib/config.cpp:244]: Probably again a null reference

File: lib/config.cpp
Comment (a-w50) [inline lib/config.cpp:271]: is null the same as json({})? in nlohman json terms?

File: lib/everest.cpp
Comment (a-w50) [inline lib/everest.cpp:333]: I really would like to change this error logging pattern with stringstream and oss, would be nice if this can be implemented with libfmt

File: lib/config.cpp
Comment (a-w50) [reply lib/config.cpp:187]: Ah probably because of the patch which populates it with a default value of {}, nice!

File: lib/config.cpp
Comment (a-w50) [reply lib/config.cpp:271]: if populated by default, then it's fine

File: schemas/manifest.json
Comment (a-w50) [inline schemas/manifest.json:5]: Good cleanup, didn't realize that this was still here

File: schemas/interface.json
Comment (a-w50) [inline schemas/interface.json:49]: Good point!
Btw, what happens if there is a default in the sub properties, is this already handled by the validate/patch mechanism?

File: lib/config.cpp
Comment (a-w50) [inline lib/config.cpp:375]: I don't really get this comment

File: lib/config.cpp
Comment (a-w50) [inline lib/config.cpp:394]: :+1: for this splitup, it makes the logic clearer and keeps the json schema more readable

File: schemas/interface.json
Comment (a-w50) [inline schemas/interface.json:32]: Why isn't the type restriction to array and string not here?
Would it be possible to use the same meta subschema for var, cmd_arguments and cmd_result?

File: everestjs/everestjs.cpp
Comment (a-w50) [inline everestjs/everestjs.cpp:96]: I'm not quite sure if this is the right way to go.  Because the on_reject isn't set now anymore, we have no chance to set a custom rejection handler.  I guess now we will only receive an exception, which we can't directly identify, i.e. from which async callback it came from - this would have been possible if we registered custom logic in the on_reject function.
So why we don't check the bool here? If it is true, we know that we're in the on_reject branch and retval should be our error object, which we can handle (that's why there was the FIXME comment) or just throw it right now.

**Review by a-w50 — CHANGES_REQUESTED** [review]
I really appreciate all the work and improvements.  Basically I only have concerns with the "Don't handle rejections, terminate with a backtrace" commit.
Concerning the list handling, I think it's an elegant solution to put all connections, whether they are single ones or multiple into a vector.  This gives a bit overhead to the single connections, as they are treated now as a vector, but I don't think this will hurt performance.  We might also use a std::variant<connection, vector<connection>> for handling this - but for now the solution is just good simple.
I also appreciate the json schema related changes!
:+1:

File: include/framework/everest.hpp
Comment (hikinggrass) [reply include/framework/everest.hpp:92]: I think this might be something we should revisit during a cleanup after implementing type-support.

File: lib/config.cpp
Comment (hikinggrass) [reply lib/config.cpp:375]: It refers to a (slightly surprising) behavior of the validator that would populate the patch with these draft-7 default values. Which in turn would lead to some very wrong non-validatable output  that we encountered during testing. Might be useful for future reference, but maybe we should clarify the intent a bit more.

Comment (a-w50) [reply lib/config.cpp:375]: which default values, if I have a non-required integer property and will validate against draft7 with patch, will this lead to the property been set to zero then, or what?

Comment (hikinggrass) [reply lib/config.cpp:375]: I have an example from the **enable command** of the **board_support_AC** interface which we found during testing:
In the main branch the validator would create the following output:
```json
{"description":"True: enabled, false: disabled.","type":"boolean"}
```

Without the changes mentioned in the comment we encountered the following output after validating with draft-7 and using the patch from the validation.
```json
{"additionalItems":true,"additionalProperties":true,"contains":true,"description":"True: enabled, false: disabled.","else":true,"if":true,"items":true,"not":true,"propertyNames":true,"readOnly":false,"then":true,"type":"boolean","uniqueItems":false}
```
The way it's implemented now prevents these (obviously wrong) additions to be patched in.

File: everestjs/utils.hpp
Comment (tmolitor-stud-tu) [reply everestjs/utils.hpp:20]: It is to get all the boost related error information (the current scope, file, line etc., basically everything you see in the log)

File: lib/config.cpp
Comment (tmolitor-stud-tu) [reply lib/config.cpp:773]: just a copy/paste hiccup, may as well be used, yes.

File: lib/config.cpp
Comment (tmolitor-stud-tu) [reply lib/config.cpp:767]: sure

File: lib/config.cpp
Comment (tmolitor-stud-tu) [reply lib/config.cpp:777]: sure

File: schemas/interface.json
Comment (tmolitor-stud-tu) [reply schemas/interface.json:49]: good question, I guess the current validation process (which includes patching the json with default values) will add defaults if not present in the data, so yes, it will be handled,but I don't know if that is desired behavior.

The better solution would be to remove all entries called `default` when loading the interface definition file, no matter their nesting level.

File: lib/config.cpp
Comment (tmolitor-stud-tu) [reply lib/config.cpp:375]: What kai said. Even worse: these additions are contradicting in itself: you never will be able to validate something having `{"not": true}` in the schema no matter what you try to validate.

File: lib/config.cpp
Comment (tmolitor-stud-tu) [reply lib/config.cpp:394]: and it prevents the "bug" discussed in the comment above :)

File: schemas/interface.json
Comment (tmolitor-stud-tu) [reply schemas/interface.json:32]: not really, at leas for the upcoming default arguments stuff already in my pipeline.
I figured having the subschemas all separated makes it easier to update them in the future and also seems a bit cleaner.

File: everestjs/everestjs.cpp
Comment (tmolitor-stud-tu) [reply everestjs/everestjs.cpp:96]: no, to the contrary: we now have a complete js backtrace pointing us directly to the line of code that caused the exception.

Comment (tmolitor-stud-tu) [reply everestjs/everestjs.cpp:96]: Second: why catch it and rethrow it manually when we could leave it to the nodejs to do exactly that?

**Comment by tmolitor-stud-tu** [comment]
I'm not sure if I can incorporate all these changes in time (e.g. in the next 1-2 weeks)

File: everestjs/everestjs.cpp
Comment (a-w50) [reply everestjs/everestjs.cpp:96]: But where would/could you catch this exception now?  If you want to have the backtrace, you could have just written
if (bool == true) throw Napi::Error
or something similar.  But I would leave it up to the user to decide on how to deal with rejected promises when using the JsExecContext helper class

**Comment by a-w50** [comment]
This PR was merged via another branch

**Pull request closed without merging**
